src/components/SwapEtherspotKlimaZap.tsx

```diff
@@ -0,0 +1,1189 @@
+    }
+  }, [web3.account])
+
+  const getBalance = (
+    currentBalances: { [ChainKey: string]: Array<TokenAmount> } | undefined,
+    chainKey: ChainKey,
+    tokenId: string,
+  ) => {
+    if (!currentBalances || !currentBalances[chainKey]) {
+      return new BigNumber(0)
+    }
+
+    const tokenBalance = currentBalances[chainKey].find(
+      (tokenAmount) => tokenAmount.address === tokenId,
+    )
+    return tokenBalance?.amount ? new BigNumber(tokenBalance?.amount) : new BigNumber(0)
+  }
+
+  useEffect(() => {
+    // merge tokens and balances
+    for (const chain of availableChains) {
+      if (!tokens[chain.key]) {
+        continue
+      }
+      for (const token of tokens[chain.key]) {
+        if (!balances || !balances[chain.key]) {
+          // balances for chain not loaded yet
+          token.amount = new BigNumber(-1)
+          token.amountRendered = undefined
+        } else {
+          // balances loaded
+          token.amount = getBalance(balances, chain.key, token.address)
+          token.amountRendered = formatTokenAmountOnly(token, token.amount)
+        }
+      }
+    }
+
+    setTokens(tokens)
+    setStateUpdate((stateUpdate) => stateUpdate + 1)
+  }, [tokens, balances, availableChains])
+
+  const hasSufficientBalance = () => {
+    if (!fromTokenAddress || !fromChainKey) {
+      return true
+    }
+
+    return depositAmount.lte(getBalance(balances, fromChainKey, fromTokenAddress))
+  }
+
+  const hasSufficientGasBalanceOnStartChain = (route?: RouteType) => {
+    if (!route) {
+      return true
+    }
+
+    const fromChain = getChainById(route.fromChainId)
+    const token = findDefaultToken(fromChain.coin, fromChain.id)
+    const balance = getBalance(balances, fromChain.key, token.address)
+
+    const requiredAmount = route.steps
+      .filter((step) => step.action.fromChainId === route.fromChainId)
+      .map(
+        (step) =>
+          step.estimate.gasCosts &&
+          step.estimate.gasCosts.length &&
+          step.estimate.gasCosts[0].amount,
+      )
+      .map((amount) => new BigNumber(amount || '0'))
+      .reduce((a, b) => a.plus(b), new BigNumber(0))
+      .shiftedBy(-18)
+    return !balance.isZero() && balance.gte(requiredAmount)
+  }
+
+  const hasSufficientGasBalanceOnCrossChain = (route?: RouteType) => {
+    if (!route) {
+      return true
+    }
+    const lastStep = route.steps[route.steps.length - 1]
+    if (!isSwapStep(lastStep)) {
+      return true
+    }
+
+    const crossChain = getChainById(lastStep.action.fromChainId)
+    const token = findDefaultToken(crossChain.coin, crossChain.id)
+    const balance = getBalance(balances, crossChain.key, token.address)
+
+    const gasEstimate =
+      lastStep.estimate.gasCosts &&
+      lastStep.estimate.gasCosts.length &&
+      lastStep.estimate.gasCosts[0].amount
+    const requiredAmount = new BigNumber(gasEstimate || 0).shiftedBy(-18)
+    return !balance.isZero() && balance.gte(requiredAmount)
+  }
```

Comment:
We should have extracted that code long time ago 😅  It's in at least 3 files now